metagraph-plugin/metagraph_katana/plugins/metagraph_katana/types.py

```diff
@@ -0,0 +1,68 @@
+import copy
+import math
+from typing import Any, Dict, List, Set
+
+import numpy as np
+from metagraph.plugins.core.types import Graph
+from metagraph.plugins.core.wrappers import GraphWrapper
+
+from katana.property_graph import PropertyGraph
+
+
+class KatanaGraph(GraphWrapper, abstract=Graph):
+    def __init__(
+        self,
+        pg_graph,
+        is_weighted=True,
+        edge_weight_prop_name="value",
+        is_directed=True,
+        node_weight_index=None,
+        node_dtype=None,
+        edge_dtype="int",
+    ):
+        super().__init__()
+        self._assert_instance(pg_graph, PropertyGraph)
+        self.value = pg_graph
+        self.is_weighted = is_weighted
+        self.edge_weight_prop_name = edge_weight_prop_name
+        self.is_directed = is_directed
+        self.node_weight_index = node_weight_index
+        self.node_dtype = node_dtype
+        self.edge_dtype = edge_dtype
+
+    def copy(self):
+        return KatanaGraph(copy.deepcopy(self.value), self.is_weighted, self.is_directed)
+
+    class TypeMixin:
+        @classmethod
+        def _compute_abstract_properties(cls, obj, props: Set[str], known_props: Dict[str, Any]) -> Dict[str, Any]:
+            ret = known_props.copy()
+            # fast props
+            for prop in {
+                "is_directed",
+                "node_type",
+                "node_dtype",
+                "edge_type",
+                "edge_dtype",
+                "edge_has_negative_weights",
+            } - ret.keys():
+                if prop == "is_directed":
+                    ret[prop] = obj.is_directed
+                if prop == "node_type":
+                    ret[prop] = "set" if obj.node_weight_index is None else "map"
+                if prop == "node_dtype":
+                    ret[prop] = None if obj.node_weight_index is None else obj.node_dtype
+                if prop == "edge_type":
+                    ret[prop] = "map" if obj.is_weighted else "set"
+                if prop == "edge_dtype":
+                    ret[prop] = obj.edge_dtype if obj.is_weighted else None
+                if prop == "edge_has_negative_weights":
+                    ret[prop] = False
```

Comment:
Let me add a TODO(pengfei): for now? Because I have not added neg-weight test cases yet.

Reply:
Like this

ret[prop] = obj.has_neg_weight #TODO(pengfei): cover the neg-weight case, and add neg-weight test cases.

Reply:
That's fine. You should also switch the default to True. Since it will prevent algorithms from running on it which don't terminate in the presence of negative edges.